nipype/interfaces/utility/tests/test_wrappers.py

```diff
@@ -50,6 +52,78 @@ def increment_array(in_array):
     wf.run()
 
 
+def test_function_inputs():
+    def f(a, b=2):
+        return a + b
+
+    def g(a, b=3):
+        return a + b
+
+    fun1 = utility.Function(function=f)
+    fun2 = utility.Function(function=f, input_names=['a', 'b'])
+    fun3 = utility.Function(function=f, input_names=['a'])
+
+    fun1.inputs.a = 1
+    fun2.inputs.a = 1
+    fun3.inputs.a = 1
+
+    assert fun1.run().outputs.out == 3
+    assert fun2.run().outputs.out == 3
+    assert fun3.run().outputs.out == 3
+
+    fun1.inputs.b = 1
+    fun2.inputs.b = 1
+
+    assert fun1.run().outputs.out == 2
+    assert fun2.run().outputs.out == 2
+
+    with pytest.raises(traits.TraitError):
+        fun3.inputs.b = 1
+
+    # Non-hidden parameters are unaffected
+    fun1.inputs.c = 1
+    fun2.inputs.c = 1
+    fun3.inputs.c = 1
+
+    # Updating the function string adds all parameters
+    fun3.inputs.function_str = getsource(g)
+    # Check we're using the new function with default b=3
+    assert fun3.run().outputs.out == 4
+    # Verify we can now set this parameter
+    fun3.inputs.b = 1
+    assert fun3.run().outputs.out == 2
+
+
+def test_function_string_reset():
+    fstr = "def f(a, b):\n\treturn a + b"
+    gstr = "def g(a, b, c):\n\treturn a + b - c"
+    hstr = "def h(b, c):\n\treturn b - c"
+
+    fun = utility.Function(function=fstr)
+    assert fun._input_names == ['a', 'b']
+    fun.inputs.a = 5
+    fun.inputs.b = 3
+    fun.inputs.c = 2  # Unused trait
```

Comment:
@effigies - i'm curious as to your thought process as to why this should not raise an `Exception`

Reply:
Because a whitelist would be a pain to maintain. IPython adds attributes to objects (or checks for them, which causes a promiscuous HasTraits to add them), and any number of environments might do something like that. Node checks for or sets num_threads, too.

It seemed to me that making sure we don't lie about the inputs to a function is more important than policing extra attributes that wouldn't be expected to change the interface's behavior.

Willing to be overruled, here, but I think we'll need a whitelist to do this right, otherwise.

Reply:
would this work?

https://docs.python.org/2.7/library/inspect.html#inspect.getargspec (in both py2 and py3)

if we get list of function variables using the function above, then:

1. for the scenario where input_names are specified, we restrict to names
2. for the scenario where all inputs are allowed, we restrict to the function variables, with any **kwargs allowing for any name.

```python
In [2]: from inspect import getargspec

In [11]: def func(a, b, c=1, **kwargs):
    ...:     pass

In [12]: getargspec(func)
Out[12]: ArgSpec(args=['a', 'b', 'c'], varargs=None, keywords='kwargs', defaults=(1,))

In [13]: def func(a, b, c=1):
    ...:     pass

In [14]: getargspec(func)
Out[14]: ArgSpec(args=['a', 'b', 'c'], varargs=None, keywords=None, defaults=(1,))
```

inside the traits object setter we could use:
```python
if name not in argspec.args and argspec.keywords is None:
    raise ValueError('{ } is not a valid input for the function.'.format(name))
```

for adding traits:
```python
if argspec.keywords is None and set(input_names) - set(argspec.args):
    raise ValueError('input_names containers undefined variable: {}'.format(set(input_names) - set(argspec.args)))

# add the input_names as traits
```

if we wanted the full possibilities, in py3 one can specifically use:

https://docs.python.org/3/library/inspect.html#inspect.getfullargspec

```
In [20]: from inspect import getfullargspec

In [22]: def func(a, b, *, c=1, **kwargs):
    ...:     pass

In [23]: getfullargspec(func)
Out[23]: FullArgSpec(args=['a', 'b'], varargs=None, varkw='kwargs', defaults=None, kwonlyargs=['c'], kwonlydefaults={'c': 1}, annotations={})
```

Reply:
I can have a look at this. At the least, it looks like a plausible replacement for the custom `_get_varnames` I'm using here.

I'm not sure how this resolves the issue of environments like IPython attempting to plop things like `_ipython_canary_method_should_not_exist_` into every object it sees. Did you have something in mind?

Reply:
i think that may be because of how `_input_names` is currently used when running the interface.  i think a lot of this can be taken care of with traits. i'll try to push a rough version on my bus ride in 30 mins.

Reply:
Oh, I do like the possibility of using annotations to make function interfaces typed. Perhaps for Nipype 2, `SimpleInterface` and `Function` could be largely merged:

```Python
def func(a: int, b: float, *, c=1: int) -> {'out1': int, 'out2': float):
    ...
```

`Function(func)` would then have:

```Python
class input_spec:
    a = traits.Int()
    b = traits.Float()
    c = traits.Int(1, usedefault=True)

class output_spec:
    out1 = traits.Int()
    out2 = traits.Float()
```

It would probably make sense to try to make `function_str` immutable once set, for simplicity.